encoding/ccf/encode.go

```diff
@@ -0,0 +1,1766 @@
+}
+
+// encodeWord64 encodes cadence.Word64 as
+// language=CDDL
+// word64-value = uint .le 18446744073709551615
+func (e *Encoder) encodeWord64(v cadence.Word64) error {
+	return e.enc.EncodeUint64(uint64(v))
+}
+
+// encodeFix64 encodes cadence.Fix64 as
+// language=CDDL
+// fix64-value = (int .ge -9223372036854775808) .le 9223372036854775807
+func (e *Encoder) encodeFix64(v cadence.Fix64) error {
+	return e.enc.EncodeInt64(int64(v))
+}
+
+// encodeUFix64 encodes cadence.UFix64 as
+// language=CDDL
+// ufix64-value = uint .le 18446744073709551615
+func (e *Encoder) encodeUFix64(v cadence.UFix64) error {
+	return e.enc.EncodeUint64(uint64(v))
+}
+
+// encodeArray encodes cadence.Array as
+// language=CDDL
+// array-value = [* value]
+func (e *Encoder) encodeArray(v cadence.Array, tids ccfTypeIDByCadenceType) error {
+	staticElementType := v.ArrayType.Element()
+
+	// Encode array head with number of array elements.
+	err := e.enc.EncodeArrayHead(uint64(len(v.Values)))
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(v.Values); i++ {
+		// Encode element as value.
+		err = e.encodeValue(v.Values[i], staticElementType, tids)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// encodeDictionary encodes cadence.Dictionary as
+// language=CDDL
+// dict-value = [* (key: value, value: value)]
+func (e *Encoder) encodeDictionary(v cadence.Dictionary, tids ccfTypeIDByCadenceType) error {
+	if len(v.Pairs) > 1 {
+		return e.encodeSortedDictionary(v, tids)
+	}
+
+	dictionaryType := v.DictionaryType.(*cadence.DictionaryType)
+	if dictionaryType == nil {
+		return fmt.Errorf("unexpected dictionary type %s", v.DictionaryType.ID())
+	}
+
+	staticKeyType := dictionaryType.KeyType
+	staticElementType := dictionaryType.ElementType
+
+	// Encode array head with array size of 2 * number of pairs.
+	err := e.enc.EncodeArrayHead(uint64(len(v.Pairs) * 2))
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(v.Pairs); i++ {
+		pair := v.Pairs[i]
+
+		// Encode dictionary key as value.
+		err = e.encodeValue(pair.Key, staticKeyType, tids)
+		if err != nil {
+			return err
+		}
+
+		// Encode dictionary value as value.
+		err = e.encodeValue(pair.Value, staticElementType, tids)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// encodeDictionary encodes cadence.Dictionary as
+// language=CDDL
+// dict-value = [* (key: value, value: value)]
+func (e *Encoder) encodeSortedDictionary(v cadence.Dictionary, tids ccfTypeIDByCadenceType) error {
+	// "Deterministic CCF Encoding Requirements" in CCF specs:
+	//
+	//   "dict-value key-value pairs MUST be sorted by key."
+
+	// Use a new buffer for sorting key value pairs.
+	buf := getBuffer()
+	defer putBuffer(buf)
+
+	// Encode and sort key value pairs.
+	sortedPairs, err := encodeAndSortKeyValuePairs(buf, v, tids)
+	if err != nil {
+		return err
+	}
+
+	// Encode array head with 2 * number of pairs.
+	err = e.enc.EncodeArrayHead(uint64(len(v.Pairs) * 2))
```

Comment:
```suggestion
	err = e.enc.EncodeArrayHead(uint64(len(v.Pairs)) * 2)
```

Reply:
Done in commit 8791232f56fcd97771f3d9a255b524e73624fc9f.